src/nvim/event/rstream.c

```diff
@@ -167,9 +169,19 @@ static void fread_idle_cb(uv_idle_t *handle)
   invoke_read_cb(stream, false);
 }
 
-static void invoke_read_cb(Stream *stream, bool eof)
+static void read_event(void **argv)
 {
+  Stream *stream = argv[0];
+  bool eof = (uintptr_t)argv[1];
   if (stream->read_cb) {
     stream->read_cb(stream, stream->buffer, stream->data, eof);
   }
+  if (eof && stream->internal_eof_cb) {
+    stream->internal_eof_cb(stream, stream->internal_data);
+  }
+}
+
+static void invoke_read_cb(Stream *stream, bool eof)
+{
+  queue_put(stream->events, read_event, 2, stream, (void *)(uintptr_t)eof);
```

Comment:
My scenario would look like this:
In the terminal you have output combined from stdout and stderr. A program generates output,
first you have some output on stdout, then output on stderr, output on stdout, output on stderr,...
The whole output should be interleaved from both streams.

Each output generates a `read_event` and they are placed in the same queue. If the queue is processed,
the first `read_event` will send the whole stdout output to the terminal (`on_job_output()` consumes the whole buffer). The next `read_event` is similar for stderr. The remaining read events do nothing because now both `RBuffer` are already empty. So the terminal would show first the stdout output and after that the stderr output.